src/Workspaces/Core/Portable/Editing/ImportAdderService.cs

```diff
@@ -142,17 +145,19 @@ private SyntaxNode GenerateNamespaceImportDeclaration(INamespaceSymbol namespace
         /// </summary>
         /// <param name="root">ref as we add simplifier annotations to nodes with explicit namespaces</param>
         /// <returns></returns>
-        private (ImmutableArray<SyntaxNode> imports, IEnumerable<INamespaceSymbol> namespaceSymbols, SyntaxNode? context) GetImportDirectivesFromSyntaxes(
+        private async ValueTask<(ImmutableArray<SyntaxNode> imports, IEnumerable<INamespaceSymbol> namespaceSymbols, SyntaxNode? context, SemanticModel model, SyntaxNode? newRoot)> GetImportDirectivesFromSyntaxesAsync(
+            Document document,
                 IEnumerable<SyntaxNode> syntaxNodes,
-                ref SyntaxNode root,
-                SemanticModel model,
+                SyntaxNode root,
                 IAddImportsService addImportsService,
                 SyntaxGenerator generator,
                 CancellationToken cancellationToken
             )
         {
             var importsToAdd = ArrayBuilder<SyntaxNode>.GetInstance();
 
+            var model = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
if this is always going to do this work (and pass back out) , i don't get why it woudln't be nicer to just have the caller compute this andpass in.

Reply:
It pushes a bunch of implementation knowledge up to the caller. This call would need to compute in advance, while the other call does not compute in advance but might compute in the implementation.

Reply:
I can implement this change if you want to see it. I felt like the additional noise didn't make it worth it.

Reply:
it just feels weird to have it not passed in, always computed, and passed out so the caller wil have it...